### ISSUE TITLE
xBIM Geometry and Revit API integration

### ISSUE THREAD

**Issue by agviegas** [xBimTeam/XbimGeometry#241]
Dear all,

I am currently trying to integrate xBIM in the Revit API. I am working in VS 2019, .Net framework 4.7.2., xBIM Geometry 5.1.288. and Revit 2020.

I aim to create IfcAnnotation objects with specific Psets that are attached to the IfcSite in the Ifc structure. I started testing the code in a Console Application and it worked perfectly (see the attached screenshot from BIMVision). Nevertheless, when I copied this code to a simple Revit External Command, it throws the following exception when the compiler reaches the `IfcStore.Open(file, editor)`:

`System.MissingMethodException: 'Method not found: 'Void Microsoft.Extensions.Logging.LoggerExtensions.LogError(Microsoft.Extensions.Logging.ILogger, System.Exception, System.String, System.Object[])'.'` 

I tried this on two different computers, and in both cases starting with a new solution with no previous NuGets packages installed. 

The External Command only contains the same code that I successfully tested in the Console application (plus the necessary code to run a Revit API External Command). You can find the code below and the solution in the following link: https://drive.google.com/open?id=1Za9jjz35DwRx1ufyBtWQdkrrqIzqsRDo 

Thank you very much in advance! :)

```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using Autodesk.Revit.DB;
using Autodesk.Revit.UI;
using Autodesk.Revit.Attributes;
using Xbim.Ifc;
using Xbim.Ifc4.MeasureResource;
using Xbim.Ifc4.PropertyResource;
using Xbim.Ifc4.Kernel;
using Xbim.Ifc4.ProductExtension;
using Xbim.Ifc4.ExternalReferenceResource;

namespace PruebasComplementoRevit
{
    [TransactionAttribute(TransactionMode.Manual)]
    public class PruebasComplementoRevit : IExternalCommand
    {
        public Result Execute(ExternalCommandData commandData, ref string message, ElementSet elements)
        {
            const string nombreDeFichero = @"C:\Users\progavie\Desktop\ANTONIO\0 REVIT API\EXPORTACION PLANTILLAS DXF\_PRUEBAS\01\ModeloArq.ifc";

            var editor = new XbimEditorCredentials
            {
                ApplicationDevelopersName = "Antonio",
                ApplicationFullName = "AddinRevit",
                ApplicationIdentifier = "asdf",
                ApplicationVersion = "4.0",


                EditorsFamilyName = "Gonzalez Viegas",
                EditorsGivenName = "Antonio",
                EditorsOrganisationName = "AGV"
            };

            using (var model = IfcStore.Open(nombreDeFichero, editor))
            {
                using (var txn = model.BeginTransaction("Transaccion"))
                {

                    // --- CREATION OF REFERENCED DWG --- 

                    //Crear el objeto de la referencia y el objeto de su asociacion al documento
                    var referenciaPlantilla = model.Instances.New<IfcDocumentReference>(p =>
                    {
                        p.Location = @"prueba\Nivel 1.dxf";
                        p.Identification = new IfcIdentifier("drawings_gen");
                        p.Description = new IfcText(" ");
                    });
                    
                    var referenciaIfcProject = model.Instances.FirstOrDefault<IfcProject>();

                    var asociacionDeReferenciaADocumento = model.Instances.New<IfcRelAssociatesDocument>(p =>
                    {
                        p.RelatingDocument = referenciaPlantilla;
                        p.RelatedObjects.Add(referenciaIfcProject);
                    }
                    );

                    // --- CREATION OF ANNOTATION OBJECTS ---

                    //Crear IfcAnnotation de vista
                    var objetoIfcAnotationDeVista = model.Instances.New<IfcAnnotation>(p => p.ObjectType = "DXFVIEW");

                    //Crear IfcAnnotation de visibilidad de  vista
                    var objetoIfcAnotationDeVisibilidadDeVista = model.Instances.New<IfcAnnotation>(p => p.ObjectType = "DXFVSVIEW");

                    //Create IfcRelContainedSpatialStructure (to IFCSite)
                    var conexionIfcAnotationConSpatialStructure = model.Instances.New<IfcRelContainedInSpatialStructure>(p =>
                    {
                        p.RelatedElements.AddRange(new[] { objetoIfcAnotationDeVista, objetoIfcAnotationDeVisibilidadDeVista });
                        p.RelatingStructure = model.Instances.FirstOrDefault<IfcSite>();
                    });

                    // --- CREATION OF PSETS --- 

                    //Create IfcRelDefinesByProperties and Visibility Psets 

                    model.Instances.New<IfcRelDefinesByProperties>(rel =>
                    {
                        rel.RelatedObjects.Add(objetoIfcAnotationDeVisibilidadDeVista);
                        rel.RelatingPropertyDefinition =

                        model.Instances.New<IfcPropertySet>(pset =>
                        {
                            pset.Name = "set_CYPE_dxf_view_visibility";
                            //Hay que añadir por separado los SingleValue y los ListValue
                            pset.HasProperties.AddRange(new[] {

                                    model.Instances.New<IfcPropertySingleValue>(p =>
                                    {
                                        p.Name = "Type";
                                        p.NominalValue = new IfcLabel("PLPLA_BUILDER"); //temporalmente, la etiqueta de IFCBUILDER
                                    })
                                });

                            pset.HasProperties.AddRange(new[] {

                                    model.Instances.New<IfcPropertySingleValue>(p =>
                                    {
                                        p.Name = "ElementGUID";
                                        p.NominalValue = new IfcIdentifier("360bWzG_bEtekNzUj6CrOd"); //temporary, to be implemented to all the Revit´s levels GUIDs
                                    })
                                });

                            pset.HasProperties.AddRange(new[] {

                                model.Instances.New<IfcPropertyListValue>(plv =>
                                    {
                                    plv.Name = "ViewIDs";
                                    plv.ListValues.Add(new IfcInteger(objetoIfcAnotationDeVista.EntityLabel));
                                    })
                                });
                        });
                    });

                    //Create IfcRelDefinesByProperties and Visibility Psets

                    model.Instances.New<IfcRelDefinesByProperties>(rel =>
                    {
                        rel.RelatedObjects.Add(objetoIfcAnotationDeVista);
                        rel.RelatingPropertyDefinition =

                        model.Instances.New<IfcPropertySet>(pset =>
                        {
                            pset.Name = "set_CYPE_dxf_view";
                          
                            pset.HasProperties.AddRange(new[] {

                                    model.Instances.New<IfcPropertySingleValue>(p =>
                                    {
                                        p.Name = "view_ID";
                                        p.NominalValue = new IfcInteger(objetoIfcAnotationDeVista.EntityLabel);
                                    }),

                                    model.Instances.New<IfcPropertySingleValue>(p =>
                                    {
                                        p.Name = "Description";
                                        p.NominalValue = new IfcText("Nivel 1");
                                    }),

                                    model.Instances.New<IfcPropertySingleValue>(p =>
                                    {
                                        p.Name = "DXF_IFCID";
                                        p.NominalValue = new IfcInteger(referenciaPlantilla.EntityLabel);
                                    }),

                                    model.Instances.New<IfcPropertySingleValue>(p =>
                                    {
                                        p.Name = "Visible";
                                        p.NominalValue = new IfcBoolean(true);
                                    }),

                                    model.Instances.New<IfcPropertySingleValue>(p =>
                                    {
                                        p.Name = "Faint";
                                        p.NominalValue = new IfcBoolean(false);
                                    }),

                                });

                            pset.HasProperties.AddRange(new[] {

                                model.Instances.New<IfcPropertyListValue>(plv => //Temporal: origen en 0, 0
                                    {
                                    plv.Name = "Origin";
                                    plv.ListValues.Add(new IfcLengthMeasure(0));
                                    plv.ListValues.Add(new IfcLengthMeasure(0));
                                    }),

                                    model.Instances.New<IfcPropertyListValue>(plv => //Temporal?
                                    {
                                    plv.Name = "TurnOrScalePoint";
                                    plv.ListValues.Add(new IfcLengthMeasure(0));
                                    plv.ListValues.Add(new IfcLengthMeasure(0));
                                    }),

                                    model.Instances.New<IfcPropertyListValue>(plv => //Temporal?
                                    {
                                    plv.Name = "Offset";
                                    plv.ListValues.Add(new IfcLengthMeasure(0));
                                    plv.ListValues.Add(new IfcLengthMeasure(0));
                                    }),
                                });

                            pset.HasProperties.AddRange(new[] {

                                    model.Instances.New<IfcPropertySingleValue>(p =>
                                    {
                                        p.Name = "XScaleFactor";
                                        p.NominalValue = new IfcReal(1);
                                    }),

                                    model.Instances.New<IfcPropertySingleValue>(p =>
                                    {
                                        p.Name = "YScaleFactor";
                                        p.NominalValue = new IfcReal(1);
                                    }),

                                    model.Instances.New<IfcPropertySingleValue>(p =>
                                    {
                                        p.Name = "TurnAngle";
                                        p.NominalValue = new IfcPlaneAngleMeasure(0);
                                    }),

                                    model.Instances.New<IfcPropertySingleValue>(p =>
                                    {
                                        p.Name = "FullView";
                                        p.NominalValue = new IfcBoolean(true);
                                    }),
                                });

                            pset.HasProperties.AddRange(new[] {

                                model.Instances.New<IfcPropertyListValue>(plv =>
                                    {
                                    plv.Name = "MinArea";
                                    plv.ListValues.Add(new IfcLengthMeasure(0));
                                    plv.ListValues.Add(new IfcLengthMeasure(0));
                                    }),

                                    model.Instances.New<IfcPropertyListValue>(plv =>
                                    {
                                    plv.Name = "MaxArea";
                                    plv.ListValues.Add(new IfcLengthMeasure(0));
                                    plv.ListValues.Add(new IfcLengthMeasure(0));
                                    }),

                                    model.Instances.New<IfcPropertyListValue>(plv =>
                                    {
                                    plv.Name = "VisibleLayersIndexes";
                                        for(int i = 0; i <= 50; i++)
                                        {
                                            plv.ListValues.Add(new IfcInteger(i));
                                        }
                                    }),
                                });
                        });
                    });


                    txn.Commit();
                }
                model.SaveAs(@"C:\Users\progavie\Desktop\prueba.ifc");

                return Result.Succeeded;

            }
        }
    }
}
```
![Screenshot of BIMVISION - successful IFC editing with a Console Application](https://user-images.githubusercontent.com/56475338/72541473-df364080-3882-11ea-9d25-22eba8a35ad3.JPG)

**Comment by andyward** [comment]
This won't be anything to do with your code - Sounds like an issue with incompatible / missing Microsoft logging versions.

(I can't make much sense of the code in Drive. A git repo would be easier....)

Rather than throw code around, my recommendation would be to create a small test project that references and makes use of Microsoft.Extensions.Loggging nuget package (against the same net472 framework). If that doesn't work it's an environmental issue. If it does, add in XBIM nugets and build the app up until you can hopefully identify the issue.

Could be a culture issue? Maybe switch to EN just to test?

Also worth checking you've got the latest VS version installed and that you've got net core etc.

**Comment by martin1cerny** [comment]
I've seen this error before. There is incompatibility between different versions of `Microsoft.Extensions.Logging` in `LoggerExtensions.LogError()` method. You should check which version of `Microsoft.Extensions.Logging` is loaded in vanilla Revit without your tool first.

**Comment by andyward** [comment]
Good call. Yes it would be good to find out what version of M.E.Logging that Revit uses.

I suspect this is exactly the same issue with #238

**Comment by agviegas** [comment]
Hello,

Thank you for your quick response. I have investigated the issue using _Process Monitor_ and found out that Revit calls `Microsoft.Extensions.Logging (version 1.1.2., located in C:\Program Files\Autodesk\Revit 2020\AddIns\PnIDModeler)`, while Xbim uses `Microsoft.Extensions.Logging (version 2.1.1.)`. It looks like both versions are not compatible, as if I try to replace the Revit´s .dll with the newer version it throws another Exception:

```
`System.TypeInitializationException: 'Exception in type initializer for class 'Xbim.Ifc.IfcStore'.'

Internal exception:
TypeLoadException: Method 'OnChange' in type 'Microsoft.Extensions.Logging.StaticFilterOptionsMonitor' of assembly 'Microsoft.Extensions.Logging, Version=2.1.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' does not have an implementation.`
```

I suspect that this can be solved using the [Appdomain´s AssemblyResolve](https://thebuildingcoder.typepad.com/blog/2014/05/rvtva3c-assembly-resolver.html), but I am not sure how to implement it in this situation. Nevertheless, I will investigate further and keep you updated.

Thanks again!

**Comment by agviegas** [comment]
Hi all,

In the end it turns out that the only solution to this .dll incompatibility issue is to use the IPC (Inter Process Communication), as pointed out by [Kim Sivonen](https://www.linkedin.com/in/kim-sivonen-0594aa2/), who kindly provided an example of this approach, and [Jeremy Tammik](https://thebuildingcoder.typepad.com/blog/2019/04/set-floor-level-and-use-ipc-for-disentanglement.html). This means separating the Xbim application and calling it from the Revit API as a process. 

Thank you very much for your help!

**Comment by SteveLockley** [comment]
Antonio I have wouond back our core dlls, to use versions 1.1.1 of microsoft logging. This may resolve, I will let you know. On other projects we are happily using the xbim libraries within Revit External events, also using separate threads, there is normally no need to run out of process. I will keep you informed

**Comment by webari** [comment]
I have a same problem.

_An exception of type 'System.MissingMethodException' occurred in TestMethodRevit.dll but was not handled in user code
Additional information: Méthode introuvable : 'Void Xbim.ModelGeometry.Scene.Xbim3DModelContext..ctor(Xbim.Common.IModel, System.String, System.String, Microsoft.Extensions.Logging.ILogger)'. occurred_

**Comment by andyward** [comment]
What version of Revit? What version of XBIM? See my notes above...

**Comment by webari** [comment]
Last version and revit 2020

**Comment by CBenghi** [comment]
Hello @webari ,
could you be more specific about the Xbim version? What version of the nuget packages are you referencing? 
If you are compiling from source, what branch and commit have you checked out?
Thanks,
Claudio

**Comment by andyward** [comment]
I've just tested loading xbim.Geometry 5.1.328 (latest nuget) into a Revit 2020 addin, and constructing a `Xbim3DModelContext` (... which is where your issue seems to be). All went fine. 

These issues are tricky to diagnose as your code is running in Revit's process and your addin is just a 'guest' alongside other addins...

It's possible that you have another addin loaded with a dependency that conflicts with the version Xbim.ModelGeometry.Scene is depending on. You will probably want to investigate an Assembly Resolver along the lines of this article:

https://forums.autodesk.com/t5/revit-api-forum/proper-way-to-handle-app-config-bindingredirects-in-revit-add-in/td-p/5692149

If you're still hitting the issue I'd suggest providing more details on versions and the code running just before the MissingMethodException. You could also try disabling all the addins to isolate the one causing the issue and then work on a fix as per the above link. 

One final tip: use the Visual Studio 'Modules' window to identify where the assemblies are being loaded from and their versions.

**Comment by agviegas** [comment]
Hi everyone! I hope you are all safe these days. 

I have tried reproducing the hello world example I posted on the [Autodesk Forum](https://forums.autodesk.com/t5/revit-api-forum/revit-api-s-integration-with-xbim-geometry-microsoft-extensions/td-p/9270358) with Revit 2021 and Xbim 5.1. (latest stable version according to the nuGet Package Manager), and it works now. 

In `C:\Program Files\Autodesk\Revit 2021\AddIns\PnIDModeler\Microsoft.Extensions.Logging.dll`, the current version is 2.0.1. It looks like this will work with any later Revit version.
